### PR TITLE
dashboard: add authority for archive experiments and events

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve?

add authority for archive experiments and events

### What is changed and how does it work?

when querying the archive experiments and events, will use `SelfSubjectAccessReviews` to judge whether the token can list the information.

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

Related changes

- [ ] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by WangXiangUSTC** [comment]
@g1eny0ung I update the HTTP request on front-end at 408ef7d, PTAL

File: pkg/apiserver/event/event.go [hunk 1]
```diff
@@ -80,12 +80,19 @@ func (s *Service) listEvents(c *gin.Context) {
...
 		FinishTimeStr:       c.Query("finishTime"),
 		ExperimentName:      c.Query("experimentName"),
-		ExperimentNamespace: c.Query("experimentNamespace"),
+		ExperimentNamespace: c.Query("namespace"),
 		UID:                 c.Query("uid"),
 		Kind:                c.Query("kind"),
 		LimitStr:            c.Query("limit"),
 	}
 
+	fmt.Println("listEvents", filter)
```
Comment (g1eny0ung) [inline pkg/apiserver/event/event.go:89]: Omit debug print.

File: ui/src/api/events.ts [hunk 2]
```diff
@@ -2,13 +2,17 @@ import { Event, EventsParams } from './events.type'
...
   http.get<Event[]>('/events', {
-    params,
+    params: {
+      namespace,
+    },
   })
 
 // Without pods field
-export const dryEvents = (params?: EventsParams) =>
+export const dryEvents = (namespace = '') =>
```
Comment (g1eny0ung) [inline ui/src/api/events.ts:13]: `namespace = ''` to `namespace = null` will be better.

Comment (g1eny0ung) [reply ui/src/api/events.ts:13]: And `EventsParams` can be removed. It will be useless.

File: pkg/apiserver/event/event.go (hunk 1 above)
Comment (WangXiangUSTC) [reply pkg/apiserver/event/event.go]: done

File: ui/src/api/events.ts (hunk 2 above)
Comment (WangXiangUSTC) [reply ui/src/api/events.ts:12]: done

**Review by g1eny0ung — APPROVED** [review]
LGTM

**Comment by WangXiangUSTC** [comment]
@STRRL PTAL

**Review by STRRL — COMMENTED** [review]
How about printing some log when `utils.CanListChaos` return `false`.

The current implementation is NOT so good for presenting authorization failed with HTTP API: 
- It seems just to return empty results(not empty array`[]`) with HTTP status 200, and the frontend needs more logic to resolve it.
- We should return `403` with authorization failed.

If this implementation is a temporary solution, just appending some logs is OK to me.

**Comment by WangXiangUSTC** [comment]
> How about printing some log when `utils.CanListChaos` return `false`.
> 
> The current implementation is NOT so good for presenting authorization failed with HTTP API:
> 
> * It seems just to return empty results(not empty array`[]`) with HTTP status 200, and the frontend needs more logic to resolve it.
> * We should return `403` with authorization failed.
> 
> If this implementation is a temporary solution, just appending some logs is OK to me.

I take the `gin.Context` as a parameter of `CanListChaos`, and will take the error info to `gin.Context`, if `CanListChaos` return false, will print the error log in dashboard @STRRL

**Comment by STRRL** [comment]
Oh, I missed it.🙈

But I think making errors in `CanListChaos` is not so good. 😰 Any operations about HTTP should be restricted in Controllers.

**Comment by cwen0** [comment]
Why not implementing a specific middleware to define this authority?

File: pkg/apiserver/archive/archive.go [hunk 3]
```diff
@@ -93,6 +93,11 @@ func (s *Service) list(c *gin.Context) {
 	name := c.Query("name")
 	ns := c.Query("namespace")
 
+	canList := utils.CanListChaos(c, ns)
+	if !canList {
+		return
```
Comment (fewdan) [inline pkg/apiserver/archive/archive.go:98]: Is it more reasonable to return some errors here, such as returning error codes (403 and so on?) ?

Comment (WangXiangUSTC) [reply pkg/apiserver/archive/archive.go:98]: yes, you are right, I will refine this logic, maybe use the middleware @cwen0 mentioned

**Comment by WangXiangUSTC** [comment]
> Why not implementing a specific middleware to define this authority?

refine this logic by using the middleware in https://github.com/chaos-mesh/chaos-mesh/pull/1261/commits/584aa1875f6c3afdacb804eec8a2b037551577c8, PTAL again @STRRL @cwen0 @g1eny0ung @fewdan

**Review by fewdan — APPROVED** [review]
LGTM

**Review by g1eny0ung — APPROVED** [review]
LGTM

**Pull request merged**